### PR TITLE
Epic 7: Avatar + Virtual Try-On

### PR DESCRIPTION
**Pull request by William-Hill**
## Summary

Implements [Epic 7 (#94)](https://github.com/William-Hill/ptso_agent/issues/94) — avatar capture/picker + Vertex AI Virtual Try-On end-to-end.

- **Backend:** new `vertex_tryon_client` wrapping `virtual-try-on-001`; cache-aware `tryon` service with sorted `item_ids` invariant; `POST /tryon/generate` route + `VirtualTryOnAgent` A2A scaffold.
- **Mobile:** new `mobile/src/lib/tryon/` data layer (queries, hooks); existing editorial UI wired up — `profile/avatar/{capture,stock,index}.tsx`, `chat/tryon.tsx`, `components/tryon/tryon-result.tsx` now perform real photo capture, DB inserts, and POST `/tryon/generate` calls. 15s client timeout falls back to `FlatLayFallback`.
- **Migration:** `00014_tryon_buckets_public.sql` flips `user-profiles` + `tryon-results` to public to match `wardrobe-images` (read via UUID-path, write/delete protected by existing RLS).
- **New dep:** `@tanstack/react-query` added to `mobile/package.json` + `QueryClientProvider` wired into `mobile/app/_layout.tsx`.

Closes #116, #117, #118, #119, #120, #121, #122, #130.

## Notes for the user (manual steps)

1. **#114 — GCP setup.** Work through `docs/runbooks/setup-vertex-tryon.md`. The branch ships a verifier (`scripts/verify_vertex_tryon.py`) that exits 0 only when creds + API + model are reachable. The runbook is the source of truth; the issue body is mildly stale.
2. **#130 — stock-model seed.** `mobile/src/lib/tryon/stock-models.ts` ships with `<paste public URL after seed>` placeholder URLs. Run `scripts/seed_stock_models.py` against ≥6 curated photos, then paste the returned public URLs into that file. Until then, the stock picker shows fallback `ImagePlaceholder` tiles and refuses to save (with a toast pointing to the runbook).
3. **`npm install` in `mobile/`** — the new `@tanstack/react-query` dep needs installing once before `pnpm dev` / `expo start`.

## Plan

`docs/superpowers/plans/2026-05-10-epic-7-avatar-tryon.md` is the executed plan.

## Test plan

- [ ] Run `scripts/verify_vertex_tryon.py` — exits 0
- [ ] Apply migration `00014_tryon_buckets_public.sql`
- [ ] Backend: `pytest -q` passes (147 tests)
- [ ] Seed ≥6 stock models, update `stock-models.ts`
- [ ] Mobile: `expo start`, log in, save a user-uploaded avatar from `/profile/avatar` → see avatar appear on `/profile`
- [ ] Save a stock-model avatar → same as above
- [ ] In chat, get an outfit recommendation → tap **See it on me** → result image fades in within ~8s
- [ ] Tap **See it on me** again on the same outfit → response shows `cache_hit: true` and image renders <500ms
- [ ] Force a timeout (drop API base URL or block network) → flat-lay fallback renders, chat surface never empty-states
- [ ] Delete the avatar from `/profile` → confirm dialog → row gone, cached `tryon_results` rows gone (cascade), storage object gone for `source='user'`

## Reviewer notes

Final code review surfaced 1 blocker (private buckets vs `getPublicUrl`) which is addressed by `00014_tryon_buckets_public.sql`. Outstanding minor items tracked as follow-ups:
- `_fetch_bytes` is serial per item — `asyncio.gather` would shave latency on cache misses with multi-item outfits.
- `useMemo` for `itemIds` derivation in `chat/tryon.tsx`.
- Mobile lacks a test runner; `mobile/src/lib/tryon/queries.ts` `deleteAvatar` URL parsing would benefit from a unit test once Jest is wired.
- `gax.InvalidArgument` collapses to HTTP 502 at the route layer; arguably a 4xx.

## Summary by CodeRabbit

* **New Features**
  * Virtual Try-On: "See it on me" generation with result preview and improved try-on flow + fallback grid
  * Avatar capture & selection: upload or choose curated stock models, pick/crop UI, and avatar deletion

* **Documentation**
  * Epic plan for Avatar + Virtual Try-On; stock model attribution and replacement guidance

* **Tests**
  * Added coverage for try-on API, service, client, and agent manifest including error/edge cases

* **Chores**
  * Public storage buckets enabled; stock seeding script and runtime tooling updates added

[Review Change Stack](https://app.coderabbit.ai/change-stack/William-Hill/ptso_agent/pull/201)

### CODE REVIEW

File: server/services/tryon.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+
+async def _cache_lookup(
+    client: Any, user_id: str, avatar_id: str, sorted_item_ids: list[str]
+) -> dict | None:
+    rows = await asyncio.to_thread(
+        lambda: client.table("tryon_results")
+        .select("id,result_url,result_path")
+        .eq("avatar_id", avatar_id)
+        .eq("item_ids", sorted_item_ids)
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/services/tryon.py:80]: Fixed in ad5f19c — added `.eq("user_id", user_id)` to `_cache_lookup` so the service-role client can't return another user's cached result. Covered by new test `test_cache_lookup_scoped_by_user_id`.

File: mobile/app/(app)/chat/tryon.tsx [hunk 2]
```diff
@@ -1,21 +1,48 @@
...
+  const tryon = useTryOn();
+
+  useEffect(() => {
+    if (avatarLoading) return;
+    if (!avatar || itemIds.length === 0) return;
+    if (tryon.isIdle) {
+      tryon.mutate({ avatarId: avatar.id, itemIds });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [avatar?.id, avatarLoading, itemIds.join(',')]);
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/app/(app)/chat/tryon.tsx]: Fixed in ad5f19c — refactored `chat/tryon.tsx` so `onTryAnother` calls a stable `fire()` callback that explicitly invokes `mutate(...)`, instead of relying on the effect to re-fire after `reset()`. Retry now actually retries on the same avatar/items.

[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/app/(app)/chat/tryon.tsx]: Fixed in ad5f19c — same fix as the Codex thread on this file: retry now fires `mutate(...)` directly via a stable `fire()` callback in `onTryAnother`, no longer dependent on the effect re-running.

File: mobile/app/(app)/profile/avatar/stock.tsx [hunk 3]
```diff
@@ -1,41 +1,64 @@
...
 export default function StockModelPickerScreen() {
   const router = useRouter();
   const insets = useSafeAreaInsets();
   const isWebWide = useIsWebWide();
   const tokens = useThemeTokens();
-  const [selectedId, setSelectedId] = useState<number>(3);
+  const { user } = useAuth();
+  const qc = useQueryClient();
+  const [selectedId, setSelectedId] = useState<string>(STOCK_MODELS[0].id);
+  const [saving, setSaving] = useState(false);
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/app/(app)/profile/avatar/stock.tsx]: Fixed in ad5f19c — `useState<string>(STOCK_MODELS[0]?.id ?? '')` so an empty list doesn't crash the screen.

File: mobile/src/components/chat/outfit-card-v2.tsx [hunk 4]
```diff
@@ -379,7 +379,12 @@ export function OutfitCardV2({
         <Button
           variant="outline"
           size="sm"
-          onPress={() => router.push('/(app)/chat/tryon')}
+          onPress={() =>
+            router.push({
+              pathname: '/(app)/chat/tryon',
+              params: { items: outfit.items.map((it) => it.id).join(',') },
+            })
+          }
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/components/chat/outfit-card-v2.tsx]: Fixed in ad5f19c — CTA now filters `outfit.items` for IDs and returns early if empty, so `items=""` cannot reach the route.

File: mobile/src/lib/tryon/queries.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  const marker = "/user-profiles/";
+  const idx = avatar.image_url.indexOf(marker);
+  if (idx < 0) return;
+  const path = avatar.image_url.slice(idx + marker.length);
+  const { error: objErr } = await supabase.storage
+    .from("user-profiles")
+    .remove([path]);
+  if (objErr) {
+    console.warn("avatar object delete failed:", objErr.message);
+  }
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/lib/tryon/queries.ts]: Fixed in ad5f19c — reordered `deleteAvatar` to remove the storage object FIRST and `throw` on error, then delete the row. Now correct under public-bucket reads (00014_tryon_buckets_public.sql) so an orphaned image can't outlive the row.

File: scripts/seed_stock_models.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            sb.storage.from_(BUCKET).upload(
+                dest,
+                f.read(),
+                {"content-type": content_type, "upsert": "true"},
+            )
+        public_url = sb.storage.from_(BUCKET).get_public_url(dest)
+        print(f"  ok  {dest} -> {public_url}")
+
+    print("\nseeded. update docs/stock-model-attribution.md with the rows above.")
+    return 0
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply scripts/seed_stock_models.py:74]: Fixed in ad5f19c — `scripts/seed_stock_models.py` now tracks uploaded/failed counts, wraps the upload in try/except, and returns 1 when any file is skipped or upload fails. Automation can now detect partial failure.

File: server/lib/vertex_tryon_client.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    if not avatar_bytes:
+        raise ValueError("avatar_bytes is required")
+
+    request = PredictRequest(endpoint=_endpoint())
+    request.instances.append(_build_instance(avatar_bytes, item_image_bytes))
+
+    def _predict() -> bytes:
+        client = PredictionServiceClient(
+            client_options={"api_endpoint": _api_endpoint()}
+        )
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/lib/vertex_tryon_client.py]: Fixed in ad5f19c — added module-level lazy `_get_client()` singleton in `vertex_tryon_client.py`. Test `autouse` fixture resets the cache between runs so each test still gets the patched class.

File: server/services/tryon.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        return TryOnResult(
+            id=cached["id"], result_url=cached["result_url"], cache_hit=True
+        )
+
+    avatar_url = await _load_avatar(client, user_id, avatar_id)
+    item_urls = await _load_item_urls(client, user_id, sorted_ids)
+
+    try:
+        avatar_bytes = await _fetch_bytes(avatar_url)
+        item_bytes = [await _fetch_bytes(u) for u in item_urls]
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/services/tryon.py]: Fixed in ad5f19c — `generate_tryon_cached` now uses `asyncio.gather(_fetch_bytes(avatar_url), *(_fetch_bytes(u) for u in item_urls))` to parallelize image fetches.

File: docs/stock-model-attribution.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+this doc whenever a photo is added, replaced, or removed.
+
+| File | Photographer | Source | License | URL |
+|---|---|---|---|---|
+| stock-01.jpg | <name> | Unsplash | Unsplash License | https://unsplash.com/photos/... |
+| stock-02.jpg | <name> | Pexels | Pexels License | https://www.pexels.com/photo/... |
+| stock-03.jpg | <name> | Burst | Burst License | https://burst.shopify.com/photos/... |
+| stock-04.jpg | ... | ... | ... | ... |
+| stock-05.jpg | ... | ... | ... | ... |
+| stock-06.jpg | ... | ... | ... | ... |
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply docs/stock-model-attribution.md:14]: Out of scope for this PR — flagged as a manual step in the PR body. Stock-model curation (#130) is user-driven; rows get filled when `scripts/seed_stock_models.py` runs against the chosen photos. Tracked under #130.

File: mobile/src/lib/tryon/stock-models.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  altText: string;
+};
+
+export const STOCK_MODELS: StockModel[] = [
+  { id: "stock-01", imageUrl: "<paste public URL after seed>", altText: "Stock model 1" },
+  { id: "stock-02", imageUrl: "<paste public URL after seed>", altText: "Stock model 2" },
+  { id: "stock-03", imageUrl: "<paste public URL after seed>", altText: "Stock model 3" },
+  { id: "stock-04", imageUrl: "<paste public URL after seed>", altText: "Stock model 4" },
+  { id: "stock-05", imageUrl: "<paste public URL after seed>", altText: "Stock model 5" },
+  { id: "stock-06", imageUrl: "<paste public URL after seed>", altText: "Stock model 6" },
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/lib/tryon/stock-models.ts:21]: Out of scope for this PR — same reason as the attribution-doc thread. Placeholder URLs are intentional until #130's manual seed step runs; the picker explicitly blocks them at runtime so a misconfigured deploy can't insert a bad URL into `user_avatars`.

File: server/lib/vertex_tryon_client.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            raise VertexTryOnError(f"vertex model not found: {exc.message}") from exc
+        except gax.PermissionDenied as exc:
+            raise VertexTryOnError(f"vertex permission denied: {exc.message}") from exc
+        except gax.InvalidArgument as exc:
+            raise VertexTryOnError(f"vertex invalid argument: {exc.message}") from exc
+        except gax.GoogleAPICallError as exc:
+            raise VertexTryOnError(f"vertex api error: {exc.message}") from exc
+        return _decode_first_prediction(response)
+
+    return await asyncio.to_thread(_predict)
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/lib/vertex_tryon_client.py:135]: Out of scope for this PR — tracked as #202 (https://github.com/William-Hill/ptso_agent/issues/202). Same architectural concern as #200 for the auto-tag side; worth solving together with a shared bounded-executor or async-client migration.

File: mobile/src/lib/tryon/use-tryon.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
+import { useMutation } from "@tanstack/react-query";
+
+import { getAccessToken } from "@/src/lib/auth";
+
+const AGENT_API_URL = process.env.EXPO_PUBLIC_AGENT_API_URL;
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/src/lib/tryon/use-tryon.ts:5]: No change — the in-mutation check is intentional. Module-load validation would crash the entire app on missing env (worse dev ergonomics for an env that's only consulted on actual try-on calls). Keeping the per-call check.

File: server/services/tryon.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+async def _load_avatar(client: Any, user_id: str, avatar_id: str) -> str:
+    rows = await asyncio.to_thread(
+        lambda: client.table("user_avatars")
+        .select("image_url")
+        .eq("id", avatar_id)
+        .eq("user_id", user_id)
+        .limit(1)
+        .execute()
+    )
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/services/tryon.py:48]: Acknowledged — same architectural concern tracked in #202 (and sibling #200 for auto-tag). Inline comments on every `asyncio.to_thread` site would create noise without action; the issue is the canonical place to discuss the trade-off.

File: mobile/app/(app)/chat/tryon.tsx [hunk 14]
```diff
@@ -1,21 +1,57 @@
...
+    () => (items ?? '').split(',').filter(Boolean),
+    [items],
+  );
+  const { data: avatar, isLoading: avatarLoading } = useActiveAvatar();
+  const tryon = useTryOn();
+
+  const fire = useCallback(() => {
+    if (!avatar || itemIds.length === 0) return;
+    tryon.mutate({ avatarId: avatar.id, itemIds });
+  }, [avatar?.id, itemIds, tryon]);
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/app/(app)/chat/tryon.tsx]: Fixed in 1648036 — `useCallback` deps now reference `tryon.mutate` (stable) instead of the whole `tryon` mutation result.

File: server/lib/vertex_tryon_client.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+def _decode_first_prediction(response) -> bytes:
+    if not response.predictions:
+        raise VertexTryOnError("vertex returned no predictions")
+    pred = response.predictions[0]
+    value = pred.struct_value if hasattr(pred, "struct_value") else pred
+    encoded = value.get("bytesBase64Encoded") if hasattr(value, "get") else value["bytesBase64Encoded"]
+    if not encoded:
+        raise VertexTryOnError("vertex prediction missing bytesBase64Encoded")
+    return base64.b64decode(encoded)
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/lib/vertex_tryon_client.py]: Fixed in 1648036 — wrapped the dict access (`KeyError`/`TypeError`/`ValueError`) and `base64.b64decode(..., validate=True)` (`binascii.Error`) in `_decode_first_prediction`, both re-raising as `VertexTryOnError`. New test `test_generate_tryon_normalizes_malformed_prediction_response` covers missing-field + bad-base64 paths.

File: server/services/tryon.py [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        lambda: client.storage.from_("tryon-results").upload(
+            path,
+            png,
+            {"content-type": "image/png"},
+        )
+    )
+    public_url = await asyncio.to_thread(
+        lambda: client.storage.from_("tryon-results").get_public_url(path)
+    )
+    return public_url, path
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply server/services/tryon.py:103]: Fixed in 1648036 — added `_delete_result_object` helper and wrapped the `tryon_results` insert in try/except so a failed insert (or post-check returning no row) removes the just-uploaded PNG from the bucket. Documented that purge_expired_tryon_results does NOT backstop these orphans (it only deletes paths referenced by a row). New test asserts `.remove()` is called when the insert raises.

File: mobile/app/(app)/chat/tryon.tsx [hunk 17]
```diff
@@ -1,21 +1,57 @@
...
 
 export default function TryOnScreen() {
   const router = useRouter();
-  const { fallback } = useLocalSearchParams<{ fallback?: string }>();
-  const isFallback = fallback === '1' || fallback === 'true';
+  const { items } = useLocalSearchParams<{ items?: string }>();
+  const itemIds = useMemo(
+    () => (items ?? '').split(',').filter(Boolean),
+    [items],
+  );
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/app/(app)/chat/tryon.tsx]: Fixed in 0925511 — `useLocalSearchParams<{ items?: string | string[] }>()` now handles both shapes via `Array.isArray(items) ? items.join(',') : items ?? ''`, and `itemIds.length === 0` now short-circuits to the `FlatLayFallback` so an empty/malformed payload no longer renders a broken idle screen.

File: mobile/app/(app)/chat/tryon.tsx [hunk 18]
```diff
@@ -1,21 +1,57 @@
...
+  const useFallback = (!avatarLoading && !avatar) || tryon.isError;
+
+  if (useFallback) {
+    return (
+      <FlatLayFallback
+        onBack={() => router.back()}
+        onTryAnother={() => {
+          tryon.reset();
+          fire();
+        }}
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply mobile/app/(app)/chat/tryon.tsx]: Fixed in 0925511 — split the fallback into three branches: `noAvatar` routes `onTryAnother` to `/(app)/profile/avatar` (the actual recovery path), `noItems` does `router.back()`, and the error case keeps `reset() + fire()`. The CTA is no longer dead in the missing-avatar state.

File: tests/test_tryon_service.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    cache_select.eq.return_value.eq.return_value.eq.return_value.gte.return_value.limit.return_value.execute.return_value.data = (
+        [cache_row] if cache_row else []
+    )
+    sb.table.return_value.insert.return_value.execute.return_value.data = [
+        {"id": inserted_id}
+    ]
+    sb.storage.from_.return_value.upload.return_value = MagicMock()
+    sb.storage.from_.return_value.get_public_url.return_value = public_url
+    return sb
+
```
[the remark this answers is not in the document]
Comment (William-Hill) [reply tests/test_tryon_service.py:26]: Fixed in 0925511 — dropped the unused `avatar_url` and `item_urls` params from `_make_supabase`. No callers were passing them.
